src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+        // tombstone sstables that are about to be expired. They most likely will have 0 overlap. But if not (due to streaming), then merge it.
+        List<SSTableStats> nearExpiryTombstones = new ArrayList<>();
+        // data sstables
+        List<SSTableStats> dataSSTables = new ArrayList<>();
+
+        OverallStats overallStats = new OverallStats();
+        sstables.forEach(s -> {
+            SSTableStats stats = getOrCompute(s);
+            collect(overallStats, stats);
+            if (s.getSSTableLevel() == SSTable.TOMBSTONE_SSTABLE_LVL)
+            {
+                // If gcGraceSeconds > 0, i.e. tombstones will linger in sstables for some time and if they are created recently,
+                // mark them as latest so that they are not compacted against data sstables just yet.
+                if (cfs.metadata.params.gcGraceSeconds > 0 && toWindowMs(s.getSSTableMetadata().maxLocalDeletionTime * 1000, gcGraceSecondsMs) >= currentGcWindowMs)
+                {
+                    latestTombstones.add(stats);
+                }
+                else if (s.getSSTableMetadata().maxLocalDeletionTime < gcBefore)
+                {
+                    compactableTombstones.add(stats);
+                }
+                else
+                {
+                    nearExpiryTombstones.add(stats);
+                }
+            }
+            else
+            {
+                dataSSTables.add(stats);
+            }
+        });
+
+        return new CategorizedSSTables(dataSSTables, compactableTombstones, nearExpiryTombstones, latestTombstones, overallStats);
+    }
+
+    static class SSTableStats
+    {
+        final SSTableReader sstable;
+        final ICardinality cardinality;
+        final long keyCount;
+        final long rowCount;
+        final TombstoneCounts tombstoneCounts;
+        final TimeWindow timeWindowMs;
+
+        public SSTableStats(SSTableReader sstable)
+        {
+            StatsMetadata meta = sstable.getSSTableMetadata();
+            this.sstable = sstable;
+            this.tombstoneCounts = new TombstoneCounts();
+            this.tombstoneCounts.partitionTombstones += meta.partitionTombstones;
+            this.tombstoneCounts.rowTombstones += meta.rowTombstones;
+            this.tombstoneCounts.rangeTombstones += meta.rangeTombstones;
+
+            ICardinality cardinality = null;
+            try
+            {
+                cardinality = SSTableReader.getCardinality(sstable);
+            }
+            catch (IOException e)
+            {
+            }
+            this.cardinality = cardinality;
+
+            if (this.cardinality != null)
+            {
+                this.keyCount = cardinality.cardinality();
+            }
+            else
+            {
+                this.keyCount = sstable.estimatedKeys();
+            }
+            this.rowCount = sstable.getTotalRows();
+            this.timeWindowMs = new TimeWindow(meta.minKey, (int) (meta.maxKey - meta.minKey));
```

Comment:
Assuming metadata is written so that maxKey is timestamp of oldest key + 1, since TimeWindow is like [start,end)

Reply:
in the metadataCollector you can see that the maxKey is calculated from timestamp in partition key + the bucket size (again present in partition key).
Its not +1. if the granularity of time bucket in PK is 250 ms, maxKey will be newest key + 250 ms.